### PR TITLE
Replace StochasticFunctions

### PR DESCRIPTION
**Pull request by colesbury**
This removes the StochasticFunctions for bernoulli, multinomial, and
normal and replaces them with functions in torch.reinforce that return
both the samples and the log-probabilities of the samples.

The current StochasticFunction implementation has a few problems: it can
be painful to use when there are multiple stochastic outputs which need
to be back-propagated through. It also requires that we store grad_fns
on Variables that have requires_grad=False in order to find stochastic
nodes.

### CODE REVIEW

**Comment by colesbury** [comment]
Here's an the corresponding change to [reinforce.py](https://gist.github.com/colesbury/d0a3398eceeb12edd40dda4215a15d31) from pytorch/examples

File: test/test_autograd.py [hunk 1]
```diff
@@ -1361,100 +1351,60 @@ def backward(self, grad_output):
...
-        samples_norm_std = torch.normal(y, stddevs)
-        z = samples_multi * 2 + 4
-        z = z + samples_multi_flat.unsqueeze(0).expand_as(samples_multi)
-        z = torch.cat([z, z], 1)
-        z = z.double()
-        z = z + samples_bernoulli + samples_norm + samples_norm_std
-        last_sample = torch.normal(z, 4)
-        z = last_sample + 2
-        self.assertFalse(z.requires_grad)
+        self.assertEqual(logpmf.data[0], math.log(0.7 if s.data[0] == 1 else 0.3))
```
Comment (apaszke) [inline test/test_autograd.py]: maybe just write it out as
```python
for val, logp, p  in zip(s.data, logpmf.data, p.data):
    self.assertEqual(logp, math.log(p if val else 1-p))
```

File: test/test_autograd.py [hunk 2]
```diff
@@ -1361,100 +1351,60 @@ def backward(self, grad_output):
...
+
+    def test_reinforce_multinomial_1d(self):
+        p = Variable(torch.Tensor([0.1, 0.2, 0.3]), requires_grad=True)
+        s, logprobs = torch.reinforce.multinomial(p)
+        self.assertEqual(s.size(), logprobs.size())
+
+        def apply_fn(p):
+            return torch.reinforce._multinomial_logpmf(s, p)
+
+        gradcheck(apply_fn, (p,), raise_exception=True)
```
Comment (apaszke) [inline test/test_autograd.py]: why can't we test `torch.reinforce.multinomial` directly? Doesn't gradcheck skip outputs that don't require grad?

File: torch/autograd/_functions/stochastic.py [hunk 3]
```diff
@@ -1,92 +1,38 @@
...
-        output_probs.add_(1e-6).reciprocal_()
-        output_probs.neg_().mul_(reward)
-        # TODO: add batched index_add
-        for i in range(probs.size(0)):
-            grad_probs[i].index_add_(0, samples[i], output_probs[i])
-        return grad_probs
-
+    @staticmethod
+    def backward(ctx, grad_output):
+        return None, None, None
```
Comment (apaszke) [inline torch/autograd/_functions/stochastic.py]: We could change this to raise an error once we start detaching Variables with `requires_grad=True`. In fact you could include that in this commit

File: torch/reinforce/__init__.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    r"""
+    Samples from the multinomial distribution characterized by `input`. Returns
+    the samples and the log-probabilities of each returned sample.
+
+    See also: :func:`torch.multinomial`
+
+    .. note::
+
+        When `num_samples` is greater than one, this function always samples
+        with replacement.
```
Comment (apaszke) [inline torch/reinforce/__init__.py]: Why? `torch.multinomial` can return multiple samples without replacement, right?

**Review by apaszke — COMMENTED** [review]
LGTM, but please move it to `torch.autograd.reinforce`. I don't think it deserves a top level package

File: torch/autograd/_functions/stochastic.py (hunk 3 above)
Comment (colesbury) [reply torch/autograd/_functions/stochastic.py]: They backward() functions actually get called for some weird reason, which is why I included the definitions.

Comment (apaszke) [reply torch/autograd/_functions/stochastic.py]: Maybe we should figure out why?

File: test/test_autograd.py (hunk 2 above)
Comment (colesbury) [reply test/test_autograd.py]: The problem is the log-probabilities depend on the drawn sample. Consider bernoulli with p=0.3:

```
bernoulli(p) = (0, log(0.7))

bernoulli(p+eps) = (0, log(0.7-eps))
bernoulli(p+eps) = (1, log(0.3+eps))
```

If the subsequent calls to the random function return a different sample, then the gradchecks would fail.

Comment (apaszke) [reply test/test_autograd.py]: Makes sense, thank you :+1:

**Comment by Kaixhin** [comment]
Just my two cents: [`tf.distributions`](https://www.tensorflow.org/api_docs/python/tf/distributions) is pretty nice. Being able to not only `sample` from a parameterised distribution, but also having access to functions such as `mean`, `cdf`, `covariance`, `prob`, `log_prob` etc. is handy. So exposing these publicly as well as using `log_prob` for `reinforce` seems like it could be useful.

File: torch/autograd/_functions/stochastic.py (hunk 3 above)
Comment (colesbury) [reply torch/autograd/_functions/stochastic.py]: I think it's because for `z = f(x, y)` if:

  x.requires_grad=False
  y.requires_grad=True

and both have `grad_fn`s defined, then z.grad_fn.next_function will include both.

I'm not particularly concerned right now for two reasons:

1. These particular functions won't last very long. Since they're simple wrappers around the torch functions they can easily be moved to C++
2. A big part of this change is to make it so that functions which don't require_grad don't have grad_fn

File: torch/reinforce/__init__.py (hunk 4 above)
Comment (colesbury) [reply torch/reinforce/__init__.py]: The probabilities are then conditional on the previously drawn samples. I'm not sure if that would affect policy-gradient methods.

**Comment by colesbury** [comment]
@Kaixhin, I think you're right. I'm torn between getting this in now and adding something like `tf.distributions` later or refactoring this to be more like `tf.distributions`.

@soumith @apaszke thoughts?

**Comment by soumith** [comment]
two thoughts on this PR:

1. removing stochastic functions is a breaking change. Now, these aren't used a lot, which makes it more acceptable to break userland than not, especially if there's huge upsides.  My question is:
  - is it essential to get a breaking change now, without deprecation? (I might have to semver the next release to 0.3, rather than 0.2.1)
  - is there no easy path to deprecation?
  - is this PR tied to the perf work? (if so, how critical is it in the overall perf).

2. `tf.distributions` style package is warranted, but in a later or parallel PR, it's a larger effort. However, if there's something that needs to be pre-designed for such a package in this PR, do think it through.

**Comment by colesbury** [comment]
Removing StochasticFunctions is important for perf because it allows us to avoid creating `grad_fn`s for variables which do not require grad. It will lets us make operations on Variables that do not require_grad nearly the same cost as operations on Tensors.

I don't see an easy path to deprecation.

I think making the next release 0.3 makes sense. The perf effort is big and a bunch of things have been deprecated, as well as a few other backwards-incompatible changes.

We could make this the minimal start of a `tf.distributions` style package. It would look something like this:
https://gist.github.com/colesbury/fdce3adff50ff92bdd05e31860e94216

It's the same functionality as this PR, but organized as classes with methods. The advantage is that when we do a full `tf.distributions` effort, we can build on it instead of having to deprecate this change. (The risk is that we get something wrong in this minimal effort that we have to fix later)

**Comment by Kaixhin** [comment]
Seems like the current approach is a breaking change, and then something like `tf.distributions` would be another breaking change. If @colesbury is able to get the changes here into a minimal version of the latter without much effort, then a 0.3 release is warranted and we simply add extra distributions from there. Nobody's expecting a fully fledged version of distributions now, and it's unlikely that we'll get it badly wrong here - worst case and we do, it won't be much of a change for users compared to the currently proposed changes.

**Comment by colesbury** [comment]
See #3165 for a minimal `tf.distributions` style package

**Comment by colesbury** [comment]
closed in favor of #3165
